ide/posts/2020-06-50-ghcide-and-nixpkgs.md

```diff
@@ -0,0 +1,321 @@
+   you can always just try to compile the package. If it works make a PR
+   against nixpkgs to remove the broken flag.
+2. Often the problem is that the package can actually build with the supplied
+   build plan but cabal doesn‘t believe us. So we can do a "jailbreak" and just
+   tell cabal to ignore the version constraints. We don‘t do this by default
+   because even if the package builds, it might now have changed semantics
+   because of a change in a dependency. So a jailbreak package should be tested
+   and reported upstream so that the cabal restrictions of that package can get
+   fixed.
+3. If those two don‘t help we can still override the build plan manually to use
+   different versions of the dependencies, not the ones provided by nixpkgs by
+   default.
+
+## How to fix ghcide in nixpkgs
+
+And the third option is what needed to be done for ghcide.
+
+There were the following problems on nixpkgs-20.03:
+
+* `hie-bios` was broken because of failing tests. Test fails during nix builds
+  are very often false positives, so I disabled the tests.
+* `ghcide` needed `regex-tdfa` and `haddock-library` newer than in the
+  stack-lts. So I just used newer versions of those two libraries. This was not
+  necessary on the haskell-updates branch because it uses a new enough stack
+  lts release.
+* `ghcide` pins the version of `haskell-lsp` and `haskell-lsp-types`. This will
+  probably be the reason why maintaining `ghcide` in nixpkgs will always be a
+  little bit of manual work because, it would have to be by chance _exactly_
+  the `haskell-lsp` version from the stack lts release, to work without manual
+  intervention.
+
+So in summary only very few lines of code were needed to get `ghcide` to work.
+If you are curious look at the commit. It
+
+* enables the generation of `haskell-lsp` and `haskell-lsp-types` 0.19.
+* uses those packages as dependencies for `ghcide`
+* disables test for `hie-bios`
+* and marks `ghcide` and `hie-bios` as unbroken.
+
+### What if I can‘t upstream my package or am stuck with an older nixpkgs version?
+
+Now I haven‘t told my story quit in order. I actually first went for another
+solution, because I needed ghcide to work with my reflex-platform i.e obelisk
+project. Obelisk uses a pinned nixpkgs version _and_ a patched ghc. So what I
+did was putting the overrides I describe above as overrides into my projects
+`default.nix`. That‘s always a nice way to first figure out how to fix a
+dependency, but of course you help a lot more people if you find a way to
+upstream the fixes into nixpkgs.  Because the nixpkgs version in
+reflex-platform was so old I had to manually create some of the packages with a
+function called `callHackageDirect`. It’s kinda the last way out, put it is
+very flexibel and should be enough to solve most dependency issues. If nothing
+else helps, create a build plan with cabal and reproduce it by hand with nix
+overrides.
+
+### ghcide in obelisk projects
```

Comment:
Can we delete this section for now? I have plans to change how we integrate and I don't want people relying on the current way. As I've discovered "experimental" means "Check out this cool new feature" to most people and has the opposite effect I was hoping for heh.

Reply:
It might be still nice to say in one sentence, "btw I did this to make ghcide work with obelisk" but not advertise that it should actually work for anyone else.